In [1]:
# -----------------------------------------------------------
# Demonstrates if there are data quality isssues in the
# provided dataset using Python 3.6
#
# Author - Aravind Padmanabhan,Texas
# Last modified - 07/06/2021
# Email - aravindp1408@gmail.com
# -----------------------------------------------------------
import pandas as pd

# Exploring Receipts data

In [2]:
df = pd.read_json('receipts.json',lines=True)
df.head()

_id  bonusPointsEarned  \
0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4  {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate  pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2                       NaN           5.0  {'$date': 1609632000000}   
3  {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}           5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus  totalSpent                    userId  
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6  
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
2             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6  
4             FINISHED         1.0  5ff1e194b6a9d73a3a9f1052

# Converting the dictionary data to a single value in '_id' column

In [3]:
df['_id'] = df['_id'].apply(pd.Series)
df.head()

_id  bonusPointsEarned  \
0  5ff1e1eb0a720f0523000575              500.0   
1  5ff1e1bb0a720f052300056b              150.0   
2  5ff1e1f10a720f052300057a                5.0   
3  5ff1e1ee0a7214ada100056f                5.0   
4  5ff1e1d20a7214ada1000561                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate  pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2                       NaN           5.0  {'$date': 1609632000000}   
3  {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}           5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus  totalSpent                    userId  
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6  
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
2             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6  
4             FINISHED         1.0  5ff1e194b6a9d73a3a9f1052

# Checking if all the records are unique

In [4]:
df['_id'].is_unique # All records are unique

True

# Verifying the datatypes

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      1119 non-null   object 
 1   bonusPointsEarned        544 non-null    float64
 2   bonusPointsEarnedReason  544 non-null    object 
 3   createDate               1119 non-null   object 
 4   dateScanned              1119 non-null   object 
 5   finishedDate             568 non-null    object 
 6   modifyDate               1119 non-null   object 
 7   pointsAwardedDate        537 non-null    object 
 8   pointsEarned             609 non-null    float64
 9   purchaseDate             671 non-null    object 
 10  purchasedItemCount       635 non-null    float64
 11  rewardsReceiptItemList   679 non-null    object 
 12  rewardsReceiptStatus     1119 non-null   object 
 13  totalSpent               684 non-null    float64
 14  userId                  

There are 15 columns and 1119 records.

Data Quality Issue 1:

There are a lot of missing values in "bonuspointsEarned" and "pointsEarned" which cannot be imputated in ths case. But with the business logic behind the points calculation, they can be imputated using regression.

# Checking the statistics of the data

In [6]:
df.describe()

bonusPointsEarned  pointsEarned  purchasedItemCount   totalSpent
count         544.000000    609.000000           635.00000   684.000000
mean          238.893382    585.962890            14.75748    77.796857
std           299.091731   1357.166947            61.13424   347.110349
min             5.000000      0.000000             0.00000     0.000000
25%             5.000000      5.000000             1.00000     1.000000
50%            45.000000    150.000000             2.00000    18.200000
75%           500.000000    750.000000             5.00000    34.960000
max           750.000000  10199.800000           689.00000  4721.950000

# Verifying if the receipts with over 600 items are valid records

In [7]:
df[df['purchasedItemCount'] > 600]

_id  bonusPointsEarned  \
469  600f39c30a7214ada2000030              750.0   
543  600f2fc80a720f0535000030              750.0   

                               bonusPointsEarnedReason  \
469  Receipt number 1 completed, bonus point schedu...   
543  Receipt number 1 completed, bonus point schedu...   

                   createDate               dateScanned  \
469  {'$date': 1611610563000}  {'$date': 1611610563000}   
543  {'$date': 1611608008000}  {'$date': 1611608008000}   

                 finishedDate                modifyDate  \
469  {'$date': 1611630363000}  {'$date': 1611630460000}   
543  {'$date': 1611612263000}  {'$date': 1611873422000}   

            pointsAwardedDate  pointsEarned              purchaseDate  \
469  {'$date': 1611630363000}        7137.2  {'$date': 1611446400000}   
543  {'$date': 1611612263000}        4944.7  {'$date': 1611446400000}   

     purchasedItemCount                             rewardsReceiptItemList  \
469               670.0  [{'brandCode': 'BETTY CROCKER', 'description':...   
543               689.0  [{'brandCode': 'BETTY CROCKER', 'description':...   

    rewardsReceiptStatus  totalSpent                    userId  
469             FINISHED     4721.95  600f35015edb782098e2ac1b  
543             FINISHED     4566.17  600f29a64329897eac239049

Verified with the given JSON that the receipts with over 600 items are valid.

# Verifying if a receipt with 10199 pointsEarned is valid

In [8]:
df[df['pointsEarned'] == 10199.8]

_id  bonusPointsEarned     bonusPointsEarnedReason  \
492  6011f3490a7214ada200010e                5.0  All-receipts receipt bonus   

                   createDate               dateScanned  \
492  {'$date': 1611789129000}  {'$date': 1611789129000}   

                 finishedDate                modifyDate  \
492  {'$date': 1611789131000}  {'$date': 1611789135000}   

            pointsAwardedDate  pointsEarned              purchaseDate  \
492  {'$date': 1611789131000}       10199.8  {'$date': 1611702729000}   

     purchasedItemCount                             rewardsReceiptItemList  \
492                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

    rewardsReceiptStatus  totalSpent                    userId  
492             FINISHED        11.0  6011f31ea4b74c18d3a8c476

Possible Data Quality Issue:

The receipt has 5 bonus points,only 2 items and total spent is 11. I did a research to find out that 1000 points equal to one dollar. So, earning 10000+ points on a total spent of 11 dollars doesn't look right. it might be an invalid record/outlier.

# Exploring Brands data

In [9]:
df1 = pd.read_json('brands.json',lines=True)
df1.head()

_id       barcode        category  \
0  {'$oid': '601ac115be37ce2ead437551'}  511111019862          Baking   
1  {'$oid': '601c5460be37ce2ead43755f'}  511111519928       Beverages   
2  {'$oid': '601ac142be37ce2ead43755d'}  511111819905          Baking   
3  {'$oid': '601ac142be37ce2ead43755a'}  511111519874          Baking   
4  {'$oid': '601ac142be37ce2ead43755e'}  511111319917  Candy & Sweets   

       categoryCode                                                cpg  \
0            BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1         BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name  topBrand                      brandCode  
0  test brand @1612366101024       0.0                            NaN  
1                  Starbucks       0.0                      STARBUCKS  
2  test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176  
3  test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051  
4  test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827

In [10]:
df1['_id'] = df1['_id'].apply(pd.Series)
df1['_id'].is_unique

True

All records are unique.

In [11]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1167 entries, 0 to 1166
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   _id           1167 non-null   object 
 1   barcode       1167 non-null   int64  
 2   category      1012 non-null   object 
 3   categoryCode  517 non-null    object 
 4   cpg           1167 non-null   object 
 5   name          1167 non-null   object 
 6   topBrand      555 non-null    float64
 7   brandCode     933 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 73.1+ KB


There are a few missing values. The dataset has 8 columns and 1167 records.

# Exploring Users data

In [12]:
df2 = pd.read_json('users.json',lines=True)
df2.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

In [13]:
df2['_id'] = df2['_id'].apply(pd.Series)
df2['_id'].is_unique

False

The records are not unique.

In [14]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   _id           495 non-null    object
 1   active        495 non-null    bool  
 2   createdDate   495 non-null    object
 3   lastLogin     433 non-null    object
 4   role          495 non-null    object
 5   signUpSource  447 non-null    object
 6   state         439 non-null    object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


There are a few missing values. The dataset has 7 columns and 495 records.

In [15]:
df2['role'].value_counts()

consumer       413
fetch-staff     82
Name: role, dtype: int64

Accroding to the description in the exercise, the role field is set to consumer. But it has a value "fetch-staff" too. There are 413 consumers and 82 fetch-staff.